README.md
# lua-resty-reqargs

Helper to Retrieve `application/x-www-form-urlencoded`, `multipart/form-data`, and `application/json` Request Arguments.

## Synopsis

```lua
local get, post, files = require "resty.reqargs"()
if not get then
    error(post)
end
-- Use get, post, and files...
```

## Installation

Just place [`reqargs.lua`](https://github.com/bungle/lua-resty-reqargs/blob/master/lib/resty/reqargs.lua)
somewhere in your `package.path`, under `resty` directory. If you are using OpenResty, the default location
would be `/usr/local/openresty/lualib/resty`.

### Using OpenResty Package Manager (opm)

```Shell
$ opm get bungle/lua-resty-reqargs
```

### Using LuaRocks

```Shell
$ luarocks install lua-resty-reqargs
```

LuaRocks repository for `lua-resty-reqargs` is located at https://luarocks.org/modules/bungle/lua-resty-reqargs.

## API

This module has only one function, and that function is loaded with require:

```lua
local reqargs = require "resty.reqargs"
```

### get, post, files regargs(options)

When you call the function (`reqargs`) you can pass it `options`. These
options override whatever you may have defined in your Nginx configuration
(or the defaults). You may use the following options:

```lua
{
    tmp_dir          = "/tmp",
    timeout          = 1000,
    chunk_size       = 4096,
    max_get_args     = 100,
    mas_post_args    = 100,
    max_line_size    = 512,
    max_file_uploads = 10
}
```

This function will return three (3) return values, and they are called
`get`, `post`,  and `files`. These are Lua tables containing the data
that was (HTTP) requested. `get` contains HTTP request GET arguments
retrieved with [ngx.req.get_uri_args](https://github.com/openresty/lua-nginx-module#ngxreqget_uri_args).
`post` contains either HTTP request POST arguments retrieved with
[ngx.req.get_post_args](https://github.com/openresty/lua-nginx-module#ngxreqget_post_args),
or in case of `application/json` (as a content type header for the request),
it will read the request body and decode the JSON, and the `post` will
then contain the decoded JSON structure presented as Lua tables. The
last return value `files` contains all the files uploaded. The `files`
return value will only contain data when there are actually files uploaded
and that the request content type is set to `multipart/form-data`. `files`
has the same structure as `get` and `post` for the keys, but the values
are presented as a Lua tables, that look like this (think about PHP's `$_FILES`):

```lua
{
    -- The name of the file upload form field (same as the key)
    name = "photo",
    -- The name of the file that the user selected for the upload
    file = "cat.jpg",
    -- The mimetype of the uploaded file
    type = "image/jpeg"
    -- The file size of the uploaded file (in bytes)
    size = 123465
    -- The location where the uploaded file was streamed
    temp = "/tmp/????"
}
```

In case of error, this function will return `nil`, `error message`.

## Nginx Configuration Variables

You can configure several aspects of `lua-resty-reqargs` directly from
the Nginx configuration, here are the configuration values that you may
use, and their default values:

```nginx
# the default is the system temp dir
set $reqargs_tmp_dir           /tmp;
# see https://github.com/openresty/lua-resty-upload
set $reqargs_timeout           1000;
# see https://github.com/openresty/lua-resty-upload
set $reqargs_chunk_size        4096;
# see https://github.com/openresty/lua-nginx-module#ngxreqget_uri_args
set $reqargs_max_get_args      100;
# see https://github.com/openresty/lua-nginx-module#ngxreqget_post_args
set $reqargs_max_post_args     100;
# see https://github.com/openresty/lua-resty-upload
set $reqargs_max_line_size     512;  
# the default is unlimited
set $reqargs_max_file_uploads  10;
```

## Changes

The changes of every release of this module is recorded in [Changes.md](https://github.com/bungle/lua-resty-reqargs/blob/master/Changes.md) file.

## License

`lua-resty-reqargs` uses two clause BSD license.

```
Copyright (c) 2015 - 2017, Aapo Talvensaari
All rights reserved.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice, this
  list of conditions and the following disclaimer in the documentation and/or
  other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES`


Changes.md
# Changelog

All notable changes to `lua-resty-reqargs` will be documented in this file.

## [1.4] - 2017-01-07
### Fixed
- Fixed issue with no options passed as reported here:
  https://groups.google.com/forum/#!topic/openresty-en/uXRXC0NbfbI

## [1.3] - 2016-09-29
### Added
- Support for the official OpenResty package manager (opm).
- Added changelog (this file).
- A lots of new documentation.

##[1.2] - 2016-08-23
### Added
- Added max_fsize option that can be used to control how large can one uploaded file be.
- Added max_files option that can be used to control how many files can be uploaded.

### Fixed
- LuaRocks etc. was using wrong directory name (renamed regargs dir to resty).

##[1.1] - 2016-08-19
### Fixed
- Files are always opened in binary mode (this affects mainly Windows users).

##[1.0] - 2016-07-06
### Added
- Initial Release.


lib/resty/reqargs.lua
local upload   = require "resty.upload"
local decode   = require "cjson.safe".decode
local tonumber = tonumber
local tmpname  = os.tmpname
local concat   = table.concat
local type     = type
local find     = string.find
local open     = io.open
local sub      = string.sub
local sep      = sub(package.config, 1, 1) or "/"
local ngx      = ngx
local req      = ngx.req
local var      = ngx.var
local body     = req.read_body
local file     = ngx.req.get_body_file
local data     = req.get_body_data
local pargs    = req.get_post_args
local uargs    = req.get_uri_args

local defaults = {
    tmp_dir          = var.reqargs_tmp_dir,
    timeout          = tonumber(var.reqargs_timeout)       or 1000,
    chunk_size       = tonumber(var.reqargs_chunk_size)    or 4096,
    max_get_args     = tonumber(var.reqargs_max_get_args)  or 100,
    max_post_args    = tonumber(var.reqargs_max_post_args) or 100,
    max_line_size    = tonumber(var.reqargs_max_line_size),
    max_file_uploads = tonumber(var.reqargs_max_file_uploads)
}

local function read(f)
    local f, e = open(f, "rb")
    if not f then
        return nil, e
    end
    local c = f:read "*a"
    f:close()
    return c
end

local function basename(s)
    local p = 1
    local i = find(s, sep, 1, true)
    while i do
        p = i + 1
        i = find(s, sep, p, true)
    end
    if p > 1 then
        s = sub(s, p)
    end
    return s
end

local function kv(r, s)
    if s == "formdata" then return end
    local e = find(s, "=", 1, true)
    if e then
        r[sub(s, 2, e - 1)] = sub(s, e + 2, #s - 1)
    else
        r[#r+1] = s
    end
end

local function parse(s)
    if not s then return nil end
    local r = {}
    local i = 1
    local b = find(s, ";", 1, true)
    while b do
        local p = sub(s, i, b - 1)
        kv(r, p)
        i = b + 1
        b = find(s, ";", i, true)
    end
    local p = sub(s, i)
    if p ~= "" then kv(r, p) end
    return r
end

return function(options)
    options = options or defaults
    local get = uargs(options.max_get_args or defaults.max_get_args)
    local post = {}
    local files = {}
    local ct = var.content_type
    if ct == nil then return get, post, files end
    if sub(ct, 1, 19) == "multipart/form-data" then
        local tmpdr = options.tmp_dir or defaults.tmp_dir
        if tmpdr and sub(tmpdr, -1) ~= sep then
            tmpdr = tmpdr .. sep
        end
        local maxfz = options.max_file_size    or defaults.max_file_size
        local maxfs = options.max_file_uploads or defaults.max_file_uploads
        local chunk = options.chunk_size       or defaults.chunk_size
        local form, e = upload:new(chunk, options.max_line_size or defaults.max_line_size)
        if not form then return nil, e end
        local h, p, f, o, s
        local u = 0
        form:set_timeout(options.timeout or defaults.timeout)
        while true do
            local t, r, e = form:read()
            if not t then return nil, e end
            if t == "header" then
                if not h then h = {} end
                if type(r) == "table" then
                    local k, v = r[1], parse(r[2])
                    if v then h[k] = v end
                end
            elseif t == "body" then
                if h then
                    local d = h["Content-Disposition"]
                    if d then
                        if d.filename then
                            if maxfz then
                                s = 0
                            end
                            f = {
                                name = d.name,
                                type = h["Content-Type"] and h["Content-Type"][1],
                                file = basename(d.filename),
                                temp = tmpdr and (tmpdr .. basename(tmpname())) or tmpname()
                            }
                            o, e = open(f.temp, "w+b")
                            if not o then return nil, e end
                            o:setvbuf("full", chunk)
                        else
                            p = { name = d.name, data = { n = 1 } }
                        end
                    end
                    h = nil
                end
                if o then
                    if maxfz then
                        s = s + #r
                        if maxfz < s then
                            o:close()
                            return nil, "The maximum size of an uploaded file exceeded."
                        end
                    end
                    if maxfs and maxfs < u + 1 then
                        o:close()
                        return nil, "The maximum number of files allowed to be uploaded simultaneously exceeded."
                    end
                    local ok, e = o:write(r)
                    if not ok then
                        o:close()
                        return nil, e
                    end
                elseif p then
                    local n = p.data.n
                    p.data[n] = r
                    p.data.n = n + 1
                end
            elseif t == "part_end" then
                if o then
                    f.size = o:seek()
                    o:close()
                    o = nil
                    if maxfs and f.size > 0 then
                        u = u + 1
                    end
                end
                local c, d
                if f then
                    c, d, f = files, f, nil
                elseif p then
                    c, d, p = post, p, nil
                end
                if c then
                    local n = d.name
                    local s = d.data and concat(d.data) or d
                    if n then
                        local z = c[n]
                        if z then
                            if z.n then
                                z.n = z.n + 1
                                z[z.n] = s
                            else
                                z = { z, s }
                                z.n = 2
                            end
                            c[n] = z
                        else
                            c[n] = s
                        end
                    else
                        c.n = c.n + 1
                        c[c.n] = s
                    end
                end
            elseif t == "eof" then
                break
            end
        end
        local t, _, e = form:read()
        if not t then return nil, e end
    elseif sub(ct, 1, 16) == "application/json" then
        body()
        local j = data()
        if j == nil then
            local f = file()
            if f ~= nil then
                j = read(f)
                if j then
                    post = decode(j) or {}
                end
            end
        else
            post = decode(j) or {}
        end
    else
        body()
        post = pargs(options.max_post_args or defaults.max_post_args)
    end
    return get, post, files
end